Spatial_Engine/Query/DominantVector.cs

```diff
@@ -0,0 +1,154 @@
+            List<double> groupedLengths = groupByNormal.Select(x => x.Select(y => y.Length()).Sum()).ToList();
+
+            //get index of biggest length, which will be dominant vector
+            int biggestLengthIndex = groupedLengths.IndexOf(groupedLengths.Max());
+
+            Vector dominantVector = groupByNormal[biggestLengthIndex].First().Normalise();
+            
+            if(!orthogonalPriority)
+                return dominantVector;
+
+            if (dominantVector.X == 0 || dominantVector.X == 1)
+                return dominantVector;
+
+            //filter grouped vectors to find only curves that are X = 0 or 1 (horizontal or vertical lines)
+            var orthogonalVectors = groupByNormal.Where(x => x.First().Normalise().X == 0 || x.First().Normalise().X == 1).ToList();
+            //then sum their total length
+            List<double> orthogonalLengths = orthogonalVectors.Select(x => x.Select(y => y.Length()).Sum()).ToList();
+            //get index of biggest length, which will be dominant vector
+            int biggestOrthogonalLengthIndex = orthogonalLengths.IndexOf(orthogonalLengths.Max());
+            
+            Vector orthogonalDominantVector = orthogonalVectors[biggestOrthogonalLengthIndex].First().Normalise();
+            if (dominantVector.IsEqual(orthogonalDominantVector))
+                return dominantVector;
+            
+            //check if length tolerance passes
+            if (orthogonalLengths.Max() < groupedLengths.Max() * orthogonalLengthTolerance)
+                return dominantVector;
+
+            return orthogonalDominantVector;
+        }
+
+        /******************************************/
+        /****            IElement2D            ****/
+        /******************************************/
+        
+        [Description("Gets the the dominant vector (orientation) of an Element2D based on its outter lines lengths.")]
+        [Input("element2D", "Element2D to evaluate.")]
+        [Input("orthogonalPriority", "Optional, if true gives priority to curves that are on the orthogonal axis.")]
+        [Input("orthogonalTolerance", "Optional, when orthogonalPriority is true it will only return orthogonal vector when its length is higher this number multiplied by the actual dominant vector lengths. For example if the dominant vector is 10 in length but the orthogonal is only 5 in length, then this number should be 0.5.")]
+        [Input("angleTolerance", "Optional, angle in radians that vectors will be considered similar. Default value in radians is approximately 5 degrees.")]
+        [Output("dominantVector", "The the dominant vector of an Element1D.")]
+        public static BH.oM.Geometry.Vector DominantVector(IElement2D element2D, bool orthogonalPriority = true, double orthogonalLengthTolerance = 0.5, double angleTolerance = 0.087)
+        {
+            IElement1D outline = BH.Engine.Geometry.Create.PolyCurve(element2D.IOutlineElements1D().Select(x =>x as ICurve));
```

Comment:
This will not work on any `IElement2D` that has outline elements different than `ICurves` (`x =>x as ICurve` will return null, `x.Geometry()` could be better).